py/dml/types.py

```diff
@@ -231,25 +256,47 @@ def sizeof(self):
         return None
 
     def cmp(self, other):
-        """Compare this type to another.
-
-        Return 0 if the types are equivalent,
-        Return NotImplemented otherwise.
-
-        The exact meaning of this is somewhat fuzzy. The
-        method is used for three purposes:
-
+        """Strict type compatibility/equality.
+
+        Return 0 if the types are run-time compatible,
+        Return NotImplemented otherwise
+
+        "Run-time compatibility" has two minimal criteria:
+        1. The C representations of the types MUST be compatible, in a C sense
+        2. A value of one type can be treated by DMLC as though it were of the
+        other type without any additional risk of undefined behavior or invalid
+        generated C.
+        For example, all trait reference types share the same C representation,
+        and so satisfy (1), but trait reference types for different traits do
+        not share vtables; trying to use a vtable for one trait with an
+        incompatible reference would result in undefined behavior, and so do
+        not satisfy (2).
+
+        The method is used for three purposes:
         1. in TPtr.canstore(), to judge whether pointer target types
         are compatible.
         2. in ctree, to compare how large values different numerical
         types can hold
-        3. when judging whether a method override is allowed, as an inaccurate
-        replacement of TPtr(self).canstore(TPtr(other))[0]
+        3. when judging whether a method override is allowed
 
         See SIMICS-9504 for further discussions.
+        """
+        return (0 if type(self) is type(other) and self.const == other.const
+                else NotImplemented)
```

Comment:
I would suggest switching signature while we are at it, e.g. `def compatible(self, other) -> bool:`
Not strictly needed within this PR, though.

Reply:
What about naming? Now we're preferring equality, should the name be `Type.eq` and `equal_types` instead of `compatible_types`?